cfg/builder/builder_entry.cc

```diff
@@ -36,13 +36,82 @@ unique_ptr<CFG> CFGBuilder::buildFor(core::Context ctx, ast::MethodDef &md) {
+                auto *join = res->freshBlock(cctx.loops, presentCont->rubyBlockId);
+                unconditionalJump(presentCont, join, *res, core::LocOffsets::none());
+                unconditionalJump(defaultCont, join, *res, core::LocOffsets::none());
+                presentCont = join;
+                defaultCont = nullptr;
+            }
+
+            // Ignore defaults for abstract methods
+            if (!isAbstract) {
+                // Only emit conditional arg loading if the arg has a default
+                if (auto *opt = ast::cast_tree<ast::OptionalArg>(argExpr)) {
+                    auto defLoc = opt->default_->loc;
+
+                    auto *presentNext = res->freshBlock(cctx.loops, presentCont->rubyBlockId);
+                    auto *defaultNext = res->freshBlock(cctx.loops, presentCont->rubyBlockId);
+
+                    auto present = cctx.newTemporary(core::Names::argPresent());
+                    synthesizeExpr(presentCont, present, core::LocOffsets::none(),
+                                   make_unique<ArgPresent>(md.symbol, i));
+                    conditionalJump(presentCont, present, presentNext, defaultNext, *res, a->loc);
+
+                    if (defaultCont) {
+                        unconditionalJump(defaultCont, defaultNext, *res, core::LocOffsets::none());
+                    }
+
+                    // Walk the default, and check the type of its final value
+                    auto result = cctx.newTemporary(core::Names::statTemp());
+                    defaultNext = walk(cctx.withTarget(result), *opt->default_, defaultNext);
+
+                    if (info.type) {
+                        auto tmp = cctx.newTemporary(core::Names::castTemp());
+                        synthesizeExpr(defaultNext, tmp, defLoc,
+                                       make_unique<Cast>(result, info.type, core::Names::let()));
+                        cctx.inWhat.minLoops[tmp.id()] = CFG::MIN_LOOP_LET;
+                    }
+
+                    // defaultNext->exprs.emplace_back(local, a->loc, make_unique<Ident>(result));
+                    synthesizeExpr(defaultNext, local, a->loc, make_unique<Ident>(result));
+
+                    presentCont = presentNext;
+                    defaultCont = defaultNext;
+                }
+            }
+
+            synthesizeExpr(presentCont, local, a->loc, make_unique<LoadArg>(md.symbol, i));
+        }
+
+        // Join the presentCont and defaultCont paths together
+        if (defaultCont) {
+            auto *join = res->freshBlock(cctx.loops, presentCont->rubyBlockId);
+            unconditionalJump(presentCont, join, *res, core::LocOffsets::none());
+            unconditionalJump(defaultCont, join, *res, core::LocOffsets::none());
+            presentCont = join;
         }
-        cont = walk(cctx.withTarget(retSym), md.rhs.get(), entry);
+
+        cont = walk(cctx.withTarget(retSym), md.rhs.get(), presentCont);
```

Comment:
This is a lot of code splatted into `buildFor`. Do you think it's worth it to factor this out into a helper with a name?

Reply:
I think that's pretty reasonable 👍